src/librustc/ty/instance.rs

```diff
@@ -290,6 +303,28 @@ impl<'tcx> Instance<'tcx> {
         result
     }
 
+    pub fn resolve_for_fn_ptr(
+        tcx: TyCtxt<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+    ) -> Option<Instance<'tcx>> {
+        debug!("resolve(def_id={:?}, substs={:?})", def_id, substs);
+        Instance::resolve(tcx, param_env, def_id, substs).map(|resolved| {
+            let resolved_def = resolved.def_id();
+            let codegen_attrs = tcx.codegen_fn_attrs(resolved_def);
+            if codegen_attrs.flags.contains(CodegenFnAttrFlags::TRACK_CALLER) {
+                debug!(" => fn pointer created for function with #[track_caller]");
```

Comment:
Maybe check that the `resolved` instance is `InstanceDef::Item`? So that you don't accidentally replace a shim with another.